capsules/extra/src/ieee802154/driver.rs

```diff
@@ -7,16 +7,59 @@
 //! Implements a userspace interface for sending and receiving IEEE 802.15.4
 //! frames. Also provides a minimal list-based interface for managing keys and
 //! known link neighbors, which is needed for 802.15.4 security.
+//!
+//! The driver functionality can be divided into three aspects: sending
+//! packets, receiving packets, and managing the 15.4 state (i.e. keys, neighbors,
+//! buffers, addressing, etc). The general design and procedure for sending and
+//! receiving is discussed below.
+//!
+//! Sending - The driver supports two modes of sending: Raw and Parse. In Raw mode,
+//! the userprocess fully forms the 15.4 frame and passes it to the driver. In Parse
+//! mode, the userprocess provides the payload and relevant metadata. From this
+//! the driver forms the 15.4 header and secures the payload. To send a packet,
+//! the userprocess issues the respective send command syscall (corresponding to
+//! raw or parse mode of sending). The 15.4 capsule will then schedule an upcall,
+//! upon completion of the transmission, to notify the process.
+//!
+//! Receiving - The driver receives 15.4 frames and passes them to the userprocess.
+//! To accomplish this, the userprocess must first `allow` a read/write ring buffer
+//! to the kernel. The kernel will then fill this buffer with received frames and
+//! schedule an upcall upon receipt of the first packet. When handling the upcall
+//! the userprocess must first `unallow` the buffer as described in section 4.4 of
+//! TRD104-syscalls. After unallowing the buffer, the userprocess must then immediately
+//! clear all pending/scheduled receive upcalls. This is done by either unsubscribing
+//! the receive upcall or subscribing a new receive upcall. Because the userprocess
+//! provides the buffer, it is responsible for adhering to this procedure. Failure
+//! to comply may result in dropped or malformed packets.
+//!
+//! The ring buffer provided by the userprocess must be of the form:
+//!
+//! | read index | write index | user_frame 0 | user_frame 1 | ... | user_frame n |
```

Comment:
```suggestion
//! ```text
//! | read index | write index | user_frame 0 | user_frame 1 | ... | user_frame n |
//! ```
```